### ISSUE TITLE
Improvements for ALIAS columns performance

### ISSUE THREAD

**Issue by ben-motiv** [ClickHouse/ClickHouse#1802]
Hi,

I am trying to create a distributed table for use with Mondrian. This requires dimensions to exist in seperate columns. Normally I would join to a seperate date dimension table but this isn't possible due to the non standard join syntax. 

This is time series data and due to the scale of it it is important that I take advantage of the partitioned nature of mergetree.

I have done a partial load of the data and find that there is a difference between:

The shard tables include a column: `month UInt8 ALIAS toMonth(datekey)`

`select count() from table where toMonth(datekey)=11 `

and

`select count() from table where month=11`

Both queries return the same correct result, however the first touches only 8 million rows (which is expected for that month) and the second touches the full 25 million in the dataset. 

It looks like an aliased column definition isn't being expanded early enough for the query optimiser?

Is there any suggestions as to how else I could have a 'month' column that takes advantage of mergetree's date partitioning?

**Comment by ben-motiv** [comment]
Partially solved by taking advantage of the partitioning key support - thanks!

However this still is a problem - I've got a lot of dictionary lookups that I would like to group by. IE primary key includes "device" and a dictionary may have a field keyed to that as "name". I want to be able to group by "name" and have it just as efficient as if group by "device".

I believe this is the purpose of "injective" attribute? I can see that it works if I directly use it in the group by clause, but it won't work if the group by uses an ALIAS'd column that expands to exactly the same thing.

Short version - should an ALIAS column be expanded before the query planner is run?

**Comment by alexey-milovidov** [comment]
Duplicate of #1647

**Comment by alexey-milovidov** [comment]
We are going to simplify ALIAS processing that will make indices to work.

**Comment by azat** [comment]
>We are going to simplify ALIAS processing that will make indices to work.

@alexey-milovidov is this in someones todo list already?

**Comment by alexey-milovidov** [comment]
No, it's not scheduled.
